### PR TITLE
Proposing PropagationPolicy Preemption API

### PR DESCRIPTION
**Pull request by RainbowMango**
**What type of PR is this?**

/kind api-change
/kind feature

**What this PR does / why we need it**:

This PR proposes the API of feature [PropagationPolicy Priority and Preemption](https://github.com/karmada-io/karmada/pull/3684). 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE(Part of feature)
```

### CODE REVIEW

File: pkg/apis/policy/v1alpha1/propagation_types.go [hunk 1]
```diff
@@ -443,6 +451,26 @@ const (
...
+type PreemptionBehavior string
+
+const (
+	// PreemptLowerPriority means that preemption is allowed as per Priority.
+	// If it is applied to a PropagationPolicy, it can preempt any resource as
+	// per Priority, regardless of whether it has been claimed by a PropagationPolicy
+	// or a ClusterPropagationPolicy, as long as it can match the rules defined
+	// in ResourceSelector.
+	// If it is applied to a ClusterPropagationPolicy, it can only preempt from
+	// ClusterPropagationPolicy, and from PropagationPolicy is not allowed.
```
Comment (Poor12) [inline pkg/apis/policy/v1alpha1/propagation_types.go:466]: I think it is necessary to add a situation where low-priority pp can preempt high-priority cpp.

Comment (Poor12) [reply pkg/apis/policy/v1alpha1/propagation_types.go:466]: PreemptLowerPriority -> PreemptLowerGrade
Grade：high-priority pp > low-priority pp > high-priority cpp > low-priority cpp

File: pkg/apis/policy/v1alpha1/propagation_types.go [hunk 2]
```diff
@@ -91,6 +92,14 @@ type PropagationSpec struct {
 	// +kubebuilder:default=0
 	Priority *int32 `json:"priority,omitempty"`
 
+	// Preemption declares the behaviors for preempting.
+	// Valid options are PreemptLowerPriority and NeverPreempt.
```
Comment (RainbowMango) [inline pkg/apis/policy/v1alpha1/propagation_types.go:96]: `PreemptLowerPriority` should be `AlwaysPreempt`.

Comment (Poor12) [inline pkg/apis/policy/v1alpha1/propagation_types.go]:
```suggestion
	// Valid options are AlwaysPreempt and NeverPreempt.
```

File: pkg/apis/policy/v1alpha1/propagation_types.go (hunk 1 above)
Comment (RainbowMango) [reply pkg/apis/policy/v1alpha1/propagation_types.go:471]: > I think it is necessary to add a situation where low-priority pp can preempt high-priority cpp.

Added in comments:
```
In addition, if a resource has already been claimed by a ClusterPropagationPolicy, the PropagationPolicy can still preempt it without considering Priority.
```

**Comment by Poor12** [comment]
/lgtm
After this PR is merged, I will update the proposal.

**Comment by RainbowMango** [comment]
I'd like to invite @Garrybest to take a look.

File: pkg/apis/policy/v1alpha1/propagation_types.go [hunk 3]
```diff
@@ -443,6 +452,30 @@ const (
...
+	// If it is applied to a PropagationPolicy, it can preempt any resource as
+	// per Priority, regardless of whether it has been claimed by a PropagationPolicy
+	// or a ClusterPropagationPolicy, as long as it can match the rules defined
+	// in ResourceSelector. In addition, if a resource has already been claimed
+	// by a ClusterPropagationPolicy, the PropagationPolicy can still preempt it
+	// without considering Priority.
+	//
+	// If it is applied to a ClusterPropagationPolicy, it can only preempt from
+	// ClusterPropagationPolicy, and from PropagationPolicy is not allowed.
+	PreemptAlways PreemptionBehavior = "AlwaysPreempt"
```
Comment (Garrybest) [inline pkg/apis/policy/v1alpha1/propagation_types.go:472]: What about just `Always` and `Never`?

Comment (RainbowMango) [reply pkg/apis/policy/v1alpha1/propagation_types.go:472]: +1, That makes sense. I can imagine how it looks in a yaml:
```
preemption: Always
```
or 
```
preemption: Never
```

**Comment by XiShanYongYe-Chang** [comment]
> /cc @chaunceyjiang for e2e failure

A webhook call timeout, we update it to 3s from 10s in #3779, and we're about to try reanalyzing it.

**Comment by RainbowMango** [comment]
@GitHubxsy Please take another look, I update a litter bit as per review comments.

**Review by whitewindmills — APPROVED** [review]
thanks, I'll work on implementing preemption of propagation policy based on this PR.
/lgtm

**Pull request merged**